test/src/test/java/org/corfudb/runtime/clients/TestClientRouter.java

```diff
@@ -242,8 +268,56 @@ public IClientRouter addClient(IClient client) {
                                                                   long epoch, Common.UuidMsg clusterId,
                                                                   CorfuMessage.PriorityLevel priority,
                                                                   boolean ignoreClusterId, boolean ignoreEpoch) {
-        // TODO(Chetan): empty stub?
-        return null;
+        // Simulate a "disconnected endpoint"
+        if (!connected) {
+            log.trace("Disconnected endpoint " + host + ":" + port);
+            throw new NetworkException("Disconnected endpoint", NodeLocator.builder()
+                    .host(host)
+                    .port(port).build());
+        }
+
+        // Set up the timer and context to measure request
+        final Timer roundTripMsgTimer = getProtoTimer(payload.getPayloadCase());
+        final Timer.Context roundTripMsgContext = MetricsUtils
+                .getConditionalContext(roundTripMsgTimer);
+
+        // Get the next request ID.
+        final long thisRequestId = requestID.getAndIncrement();
+        Common.UuidMsg protoClientId = CorfuProtocolCommon.getUuidMsg(clientID);
+
+        // Set the base fields for this message.
+        CorfuMessage.HeaderMsg header = CorfuProtocolMessage.getHeaderMsg(thisRequestId, priority,
+                epoch, clusterId, protoClientId, ignoreClusterId, ignoreEpoch);
+        CorfuMessage.RequestMsg request = CorfuMessage.RequestMsg.newBuilder().setHeader(header)
+                .setPayload(payload).build();
+
+        // Generate a future and put it in the completion table.
+        final CompletableFuture<T> cf = new CompletableFuture<>();
+        outstandingRequests.put(thisRequestId, cf);
+
+        // TODO: TestRule and TestServerRouter
+        // Evaluate rules.
+        if (rules.stream().allMatch(x -> x.evaluate(request, this))) {
+            // Write the message out to the channel
+            log.trace(Thread.currentThread().getId() + ":Sent request: {}", request);
+            // No need to simulate serialization for Protobuf
+            serverRouter.sendServerMessage(request, channelContext);
```

Comment:
Why not simulate serialization? Here is my reasoning, feel free to add on.

1. For the time-being, we add request and response marker byte into the buffer, its better to simulate/test that.
2. The serialization is simulated on the server side. Either we can remove there, or add it here just to keep the consistency.
3. Test suites should generally mock the actual behaviours of the main code to cover all the corner cases. The main idea anyway is to catch any possible error during the testing phase itself.